File: packages/agentmesh-orchestrator/src/agentmesh_orchestrator/models/task.py
"""
AgentMesh Task 数据模型

基于 task.yaml 结构和 gui.md API 规范定义的 Pydantic 模型。
"""

from datetime import datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel, Field


class TaskState(str, Enum):
    """任务状态"""
    CREATED = "created"
    WORKING = "working"
    GATE_BLOCKED = "gate.blocked"
    COMPLETED = "completed"
    FAILED = "failed"


class TaskTopology(str, Enum):
    """协作拓扑模式"""
    SWARM = "swarm"      # 蜂群模式 - 并发 Debug / 信息搜集
    SQUAD = "squad"      # 小队模式 - 全栈开发


class AgentInstanceState(str, Enum):
    """Agent 实例状态"""
    CREATED = "created"
    ACTIVE = "active"
    AWAITING = "awaiting"
    DORMANT = "dormant"
    COMPLETED = "completed"
    FAILED = "failed"


class GateState(str, Enum):
    """Gate 状态"""
    OPEN = "open"
    BLOCKED = "blocked"
    APPROVED = "approved"
    REJECTED = "rejected"


# =============================================================================
# 子结构模型
# =============================================================================

class Milestone(BaseModel):
    """里程碑"""
    id: str
    title: str
    state: str = "pending"  # pending / in_progress / done


class AgentInstance(BaseModel):
    """Agent 实例配置"""
    instance: str           # 实例 ID, 如 "architect-1"
    agent: str              # Agent spec 名称, 如 "architect"
    state: AgentInstanceState = AgentInstanceState.CREATED
    assigned_milestone: str | None = Field(None, alias="assignedMilestone")
    skills: list[str] = Field(default_factory=list)

    model_config = {"populate_by_name": True}


class Gate(BaseModel):
    """人工介入点"""
    id: str
    type: str               # applyPatch / execCommand / custom
    state: GateState = GateState.OPEN
    reason: str = ""
    agent_instance: str | None = Field(None, alias="agentInstance")
    created_at: datetime | None = Field(None, alias="createdAt")
    resolved_at: datetime | None = Field(None, alias="resolvedAt")
    decision: str | None = None  # approved / rejected
    comment: str | None = None

    model_config = {"populate_by_name": True}


class TaskConfig(BaseModel):
    """任务配置"""
    max_concurrent_agents: int = Field(3, alias="maxConcurrentAgents")
    timeout_seconds: int = Field(3600, alias="timeoutSeconds")
    auto_approve: bool = Field(False, alias="autoApprove")

    model_config = {"populate_by_name": True}


# =============================================================================
# Task 完整模型 (对应 task.yaml)
# =============================================================================

class Task(BaseModel):
    """完整的任务模型 (对应 task.yaml)"""
    id: str
    title: str
    description: str = ""
    topology: TaskTopology = TaskTopology.SQUAD
    state: TaskState = TaskState.CREATED
    created_at: datetime = Field(default_factory=datetime.utcnow, alias="createdAt")
    updated_at: datetime = Field(default_factory=datetime.utcnow, alias="updatedAt")
    milestones: list[Milestone] = Field(default_factory=list)
    roster: list[AgentInstance] = Field(default_factory=list)
    gates: list[Gate] = Field(default_factory=list)
    config: TaskConfig = Field(default_factory=TaskConfig)

    model_config = {"populate_by_name": True}


# =============================================================================
# API 响应模型
# =============================================================================

class TaskSummary(BaseModel):
    """任务摘要 (用于列表展示)"""
    id: str
    title: str
    state: TaskState
    topology: TaskTopology
    created_at: datetime = Field(..., alias="createdAt")
    updated_at: datetime = Field(..., alias="updatedAt")
    agent_count: int = Field(0, alias="agentCount")
    active_gates: int = Field(0, alias="activeGates")

    model_config = {"populate_by_name": True, "by_alias": True}


class TaskDetail(BaseModel):
    """任务详情 (用于详情页)"""
    id: str
    title: str
    description: str
    topology: TaskTopology
    state: TaskState
    created_at: datetime = Field(..., alias="createdAt")
    updated_at: datetime = Field(..., alias="updatedAt")
    milestones: list[Milestone]
    roster: list[AgentInstance]
    gates: list[Gate]
    config: TaskConfig

    model_config = {"populate_by_name": True, "by_alias": True}


class TaskEvent(BaseModel):
    """任务事件 (来自 events.jsonl)"""
    ts: datetime
    type: str
    task_id: str = Field(..., alias="taskId")
    agent_instance: str | None = Field(None, alias="agentInstance")
    turn_id: str | None = Field(None, alias="turnId")
    payload: dict[str, Any] = Field(default_factory=dict)
    by: str | None = None
    path: str | None = None

    model_config = {"populate_by_name": True, "by_alias": True}


# =============================================================================
# API 请求模型
# =============================================================================

class CreateTaskRequest(BaseModel):
    """创建任务请求"""
    title: str
    description: str = ""
    topology: TaskTopology = TaskTopology.SQUAD
    milestones: list[Milestone] = Field(default_factory=list)
    roster: list[AgentInstance] = Field(default_factory=list)
    config: TaskConfig | None = None


class CreateTaskResponse(BaseModel):
    """创建任务响应"""
    id: str
    message: str = "Task created successfully"

File: crates/agentmesh-core/src/task.rs
use chrono::DateTime;
use chrono::Utc;
use serde::Deserialize;
use serde::Serialize;

fn now() -> DateTime<Utc> {
    Utc::now()
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "kebab-case")]
pub enum TaskTopology {
    Swarm,
    Squad,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "kebab-case")]
pub enum TaskState {
    Created,
    Working,
    #[serde(alias = "gate.blocked")]
    InputRequired,
    Completed,
    Failed,
    #[serde(alias = "cancelled")]
    Canceled,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum MilestoneState {
    Pending,
    #[serde(alias = "in_progress")]
    Working,
    Done,
    Blocked,
}

impl Default for MilestoneState {
    fn default() -> Self {
        Self::Pending
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct Milestone {
    pub id: String,
    pub title: String,
    #[serde(default)]
    pub state: MilestoneState,
    #[serde(default)]
    pub depends_on: Vec<String>,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AgentInstanceState {
    #[serde(alias = "created")]
    Pending,
    Active,
    Awaiting,
    Dormant,
    Completed,
    Failed,
}

impl Default for AgentInstanceState {
    fn default() -> Self {
        Self::Pending
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct AgentInstance {
    pub instance: String,
    pub agent: String,
    #[serde(default)]
    pub state: AgentInstanceState,
    pub assigned_milestone: Option<String>,
    #[serde(default)]
    pub skills: Vec<String>,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "kebab-case")]
pub enum GateType {
    HumanApproval,
    AutoCheck,
    MilestoneGate,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum GateState {
    Open,
    Blocked,
    Approved,
    Rejected,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct Gate {
    pub id: String,
    #[serde(rename = "type")]
    pub gate_type: GateType,
    pub state: GateState,
    #[serde(default)]
    pub reason: String,
    pub instructions_ref: Option<String>,
    pub blocked_at: Option<DateTime<Utc>>,
    pub resolved_at: Option<DateTime<Utc>>,
    pub resolved_by: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct TaskConfig {
    pub max_concurrent_agents: u32,
    pub timeout_seconds: u32,
    pub auto_approve: bool,
}

impl Default for TaskConfig {
    fn default() -> Self {
        Self {
            max_concurrent_agents: 3,
            timeout_seconds: 3600,
            auto_approve: false,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct TaskFile {
    pub id: String,
    pub title: String,
    #[serde(default)]
    pub description: String,
    pub topology: TaskTopology,
    pub state: TaskState,
    #[serde(default = "now")]
    pub created_at: DateTime<Utc>,
    #[serde(default = "now")]
    pub updated_at: DateTime<Utc>,
    #[serde(default)]
    pub milestones: Vec<Milestone>,
    #[serde(default)]
    pub roster: Vec<AgentInstance>,
    #[serde(default)]
    pub gates: Vec<Gate>,
    #[serde(default)]
    pub config: TaskConfig,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct CreateTaskRequest {
    pub title: String,
    #[serde(default)]
    pub description: String,
    pub topology: TaskTopology,
    #[serde(default)]
    pub milestones: Vec<Milestone>,
    #[serde(default)]
    pub roster: Vec<AgentInstance>,
    pub config: Option<TaskConfig>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct CreateTaskResponse {
    pub id: String,
    pub message: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct TaskEvent {
    pub ts: DateTime<Utc>,
    #[serde(rename = "type")]
    pub event_type: String,
    pub task_id: String,
    pub agent_instance: Option<String>,
    pub turn_id: Option<String>,
    #[serde(default)]
    pub payload: serde_json::Value,
    pub by: Option<String>,
    pub path: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ClusterStatus {
    pub orchestrator: String,
    pub codex_adapter: String,
    pub active_agents: u32,
    pub max_agents: u32,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn agent_instance_state_accepts_legacy_created() {
        let state: AgentInstanceState = serde_yaml::from_str("created").unwrap();
        assert_eq!(state, AgentInstanceState::Pending);
    }

    #[test]
    fn milestone_state_accepts_legacy_in_progress() {
        let state: MilestoneState = serde_yaml::from_str("in_progress").unwrap();
        assert_eq!(state, MilestoneState::Working);
    }

    #[test]
    fn task_state_accepts_legacy_gate_blocked() {
        let state: TaskState = serde_yaml::from_str("gate.blocked").unwrap();
        assert_eq!(state, TaskState::InputRequired);
    }

    #[test]
    fn task_state_accepts_legacy_cancelled() {
        let state: TaskState = serde_yaml::from_str("cancelled").unwrap();
        assert_eq!(state, TaskState::Canceled);
    }
}
